integrations/access/common/access_monitoring_rules.go

```diff
@@ -0,0 +1,177 @@
+		ruleAppliesCallback:        conf.RuleAppliesCallback,
+		fetchRecipientCallback:     conf.FetchRecipientCallback,
+		matchAccessRequestCallback: conf.MatchAccessRequestCallback,
+	}
+}
+
+func (amrh *AccessMonitoringRuleHandler) InitAccessMonitoringRulesCache(ctx context.Context) error {
+	accessMonitoringRules, err := amrh.getAllAccessMonitoringRules(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	amrh.accessMonitoringRules.Lock()
+	defer amrh.accessMonitoringRules.Unlock()
+	for _, amr := range accessMonitoringRules {
+		if !amrh.ruleAppliesCallback(amr) {
+			continue
+		}
+		amrh.accessMonitoringRules.rules[amr.GetMetadata().Name] = amr
+	}
+	return nil
+}
+
+func (amrh *AccessMonitoringRuleHandler) HandleAccessMonitoringRule(ctx context.Context, event types.Event) error {
+	if kind := event.Resource.GetKind(); kind != types.KindAccessMonitoringRule {
+		return trace.BadParameter("expected %s resource kind, got %s", types.KindAccessMonitoringRule, kind)
+	}
+
+	amrh.accessMonitoringRules.Lock()
+	defer amrh.accessMonitoringRules.Unlock()
+	switch op := event.Type; op {
+	case types.OpPut:
+		e, ok := event.Resource.(types.Resource153Unwrapper)
+		if !ok {
+			return trace.BadParameter("expected Resource153Unwrapper resource type, got %T", event.Resource)
+		}
+		req, ok := e.Unwrap().(*accessmonitoringrulesv1.AccessMonitoringRule)
+		if !ok {
+			return trace.BadParameter("expected AccessMonitoringRule resource type, got %T", event.Resource)
+		}
+
+		// In the event an existing rule no longer applies we must remove it.
+		if !amrh.ruleAppliesCallback(req) {
+			delete(amrh.accessMonitoringRules.rules, event.Resource.GetName())
+			return nil
+		}
+		amrh.accessMonitoringRules.rules[req.Metadata.Name] = req
+		return nil
+	case types.OpDelete:
+		delete(amrh.accessMonitoringRules.rules, event.Resource.GetName())
+		return nil
+	default:
+		return trace.BadParameter("unexpected event operation %s", op)
+	}
+}
+
+func (amrh *AccessMonitoringRuleHandler) RecipientsFromAccessMonitoringRules(ctx context.Context, req types.AccessRequest) *RecipientSet {
+	log := logger.Get(ctx)
+	recipientSet := NewRecipientSet()
+
+	for _, rule := range amrh.getAccessMonitoringRules() {
+		match, err := amrh.matchAccessRequestCallback(rule.Spec.Condition, req)
+		if err != nil {
+			log.WithError(err).WithField("rule", rule.Metadata.Name).
+				Warn("Failed to parse access monitoring notification rule")
+		}
+		if !match {
+			continue
+		}
+		for _, recipient := range rule.Spec.Notification.Recipients {
+			rec, err := amrh.fetchRecipientCallback(ctx, recipient)
+			if err != nil {
+				log.WithError(err).Warn("Failed to fetch plugin recipients based on Access moniotring rule recipients")
+				continue
+			}
+			recipientSet.Add(*rec)
+		}
+	}
+	return &recipientSet
+}
+
+func (amrh *AccessMonitoringRuleHandler) getAllAccessMonitoringRules(ctx context.Context) ([]*accessmonitoringrulesv1.AccessMonitoringRule, error) {
```

Comment:
I'm always afraid of these `getAll` methods.
It seems we do some filtering to this list. Can we filter them as soon as we fetch a page? and only append the "good" ones?

Reply:
I'm unsure what extra filtering you mean here.

Reply:
```go
	for _, amr := range accessMonitoringRules {
		if !amrh.ruleAppliesCallback(amr) {
			continue
		}
		amrh.accessMonitoringRules.rules[amr.GetMetadata().Name] = amr
	}
```

Reply:
I can add this in but the rule applies callback just checks for pluginType and that the resource the rule is about is an access request which is already being filtered by the ListAccessMonitoringRulesWithFilter. I suppose an extra check never hurts and someone may want to add extra checks to their ruleAppliesCallback

Reply:
Oh wait nevermind i see what you mean, did not realize this was happening straight after the call anyway

Reply:
Thank you